src/AppInstallerCLICore/Workflows/DependenciesFlow.cpp

```diff
@@ -82,7 +86,142 @@ namespace AppInstaller::CLI::Workflow
+            installer->Dependencies.ApplyToType(DependencyType::Package, [&](Dependency dependency)
+                {
+                    toCheck.push_back(dependency);
+                    dependencyGraph[dependency.Id] = std::vector<Dependency>();
+                    dependencyGraph[installer->ProductId].push_back(dependency);
+
+                });
+        } // TODO fail otherwise
+
+        for (const auto& dependencyNode : toCheck)
+        {
+            // search the package source+installed to see if the dep exists
+            SearchRequest searchRequest;
+            searchRequest.Filters.emplace_back(PackageMatchFilter(PackageMatchField::Id, MatchType::CaseInsensitive, dependencyNode.Id));
+            const auto& matches = source->Search(searchRequest).Matches;
+
+            if (!matches.empty())
+            {
+                const auto& match = matches.at(0); // What to do if there's more than one? TODO should not happen (report to the user)
+                const auto& package = match.Package;
+                if (!package->GetInstalledVersion())
+                {
+                    const auto& packageVersion = package->GetLatestAvailableVersion(); //TODO check availability before using
+                    //how to choose best installer, should I use SelectInstaller or not?
+                    const auto& packageVersionManifest = packageVersion->GetManifest();
+                    const auto& matchInstaller = packageVersionManifest.Installers.at(0); // fail if there's no installer?
+                    const auto& matchDependencies = matchInstaller.Dependencies;
+                    auto matchId = matchInstaller.ProductId; // ProductId is the same Id as the one used by Dependencies?
+
+                    // TODO save installers for later maybe?
+                    matchDependencies.ApplyToType(DependencyType::Package, [&](Dependency dependency)
+                        {
+                            dependencyGraph[matchId].push_back(dependency); 
+
+                            auto search = dependencyGraph.find(dependency.Id);
+                            if (search == dependencyGraph.end()) // if not found
+                            {
+                                    toCheck.push_back(dependency);
+                                    dependencyGraph[dependency.Id] = std::vector<Dependency>();
+                            } 
+                            else
+                            {
+                                // we only need to check for loops if the dependency already existed, right?
+                                // should we have an inverse map? i.e., < id, packages that depend on this one >
+                                // or should we check for loops only once (when we have all deps in the graph)
+                                if (graphHasLoop(dependencyGraph))

```

Comment:
Isn't the definition of `graphHasLoop` that we see a dependency that we have already seen, aka this block of code is being executed?

Reply:
No, we can get a dependency that we already saw but there's no loop between them. Example:
   A depends on B and C; C depends on B
B is already on the graph but there's no loop, in fact we only need to make sure we install B first.